### DISCUSSION TITLE
[firebase_auth] Example for Custom OAuth provider flow for Microsoft Active Directory

### DISCUSSION THREAD

**Discussion by jthegedus** [firebase/flutterfire#3522] [Q&A]
#1107 was asking about Microsoft OAuth & Active Directory support since Firebase added the feature back in 2019.

A [recent update](https://github.com/FirebaseExtended/flutterfire/issues/1107#issuecomment-680070414) in that thread from @Salakar mentioned the roadmap and rework (kudos everyone) and then says:

>this feature is should now be supported through the custom OAuth provider flow

Looking at the [FlutterFire Auth documentation](https://firebase.flutter.dev/docs/auth/overview) there doesn't appear to be a reference to Custom Auth providers. And the Social Auth section containing Google, FB, Apple, Twitter and GitHub would indicate that Microsoft would follow as a direct implementation.

Can someone shed light on:
- the current state of the docs (are they just behind the implementation?)
- if there is an ETA on Microsoft AD support?
- if there is an ETA or WIP documentation for Custom OAuth provider flows?

Thanks (I'm new to Dart/Flutter so excuse if custom OAuth is a simple solution, I cannot find any ref to it)

**Comment by paurakhsharma** [comment]
I have the same question @jthegedus, did you find any solution?

**Comment by jthegedus** [comment]
AD is supported via the generic `OAuthProvider()`. It was pretty easy to implement, but IMO the docs should have at least one example to reduce the confusion and demonstrate the FlutterFire Auth libraries completeness.

This is how I did it with FlutterFire Auth for a Web project:

```dart
FirebaseApp myApp = Firebase.app('my_app');
FirebaseAuth auth = FirebaseAuth.instanceFor(app: myApp);

OAuthProvider provider = OAuthProvider('microsoft.com');
provider.setCustomParameters({'tenant': 'your_tenant_id_here'});
await auth.signInWithPopup(provider);
```

[EDIT]: WEB ONLY

**Reply by IDONTSUDO** [reply]
Thx!111

**Reply by iharshb** [reply]
What to do for App ?

**Reply by paurakhsharma** [reply]
@iharshb Try this: https://stackoverflow.com/a/64218952/8468530

**Comment by paurakhsharma** [comment]
@jthegedus Thank you so much for the response, but I believe this only works for a web project. 
Any suggestions on how I can do this for android and iOS?

**Reply by jthegedus** [reply]
I need to do this myself over the next few days, so will certainly have an update. Again, I wish the docs had this in them

**Reply by paurakhsharma** [reply]
Thank you so much, currently, I am trying to get the `accessToken` and `idToken` using https://pub.dev/packages/aad_oauth

and then use 

```dart
final AadOAuth oauth = new AadOAuth(config);
await oauth.login();
String accessToken = await oauth.getAccessToken();
String idToken = await oauth.getIdToken();

OAuthProvider provider = OAuthProvider('microsoft.com');

OAuthCredential credential = provider.credential(accessToken: accessToken, idToken: idToken);

await FirebaseAuth.instance.signInWithCredential(credential);
```

But, my problem is `idToken` is always `null` for me. I am able to get the `accessToken`.

So, I get this error. 
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: [firebase_auth/unknown] Attempt to invoke virtual method 'java.lang.Class java.lang.Object.getClass()' on a null object reference
```

If I can solve this I will update here, else waiting for your update.

**Reply by jthegedus** [reply]
Firebase Auth OAuth works okay:

**Links**:

- https://pub.dev/packages/firebase_auth_oauth
- https://github.com/amrfarid140/firebase_auth_oauth

**Example**:

```yaml
# pubspec.yaml
dependencies:
  flutter:
    sdk: flutter
  firebase_auth: ^0.18.0+1
  firebase_core: ^0.5.0
  firebase_auth_oauth: ^0.2.0
```

```dart
// login.dart
import 'package:firebase_auth_oauth/firebase_auth_oauth.dart';
...
User user = await FirebaseAuthOAuth().openSignInFlow(
  "microsoft.com", ["email"], {'tenant': 'your_tenant_id_here'});
```

**Caveats**:

- it doesn't appear to integrate with iOS shared Keychain or Android shared MSAL (I may be wrong)
- BSD-3 license
- requires creating OAuth2/API key for your app - https://developers.google.com/android/guides/client-auth

I also have concerns around keeping the Firebase dependencies of my app with `firebase_auth_oauth` aligned.

---

I am still testing and pursuing a solution like you shared as I require iOS Keychain integration etc.

**Reply by paurakhsharma** [reply]
Thank you so much, I will give it a shot.

**Reply by jthegedus** [reply]
@paurakhsharma In your example, it seems to get back an `id_token` from Microsoft you need to provide the `scope` of `openid` - https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#send-the-sign-in-request

So perhaps your example will work with the scopes of both `email` and `openid`? Please try and let me know.

**Reply by paurakhsharma** [reply]
Yeah, actually I get the `id_token` when the `scope` is `email` and `openid`.

But still this line: `await FirebaseAuth.instance.signInWithCredential(credential);`

Fails with error: 

```
[ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: [firebase_auth/unknown] com.google.firebase.FirebaseException: An internal error has occurred. [ INVALID_CREDENTIAL_OR_PROVIDER_ID:Invalid IdP response/credential: http://localhost?id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6ImppYk5ia0ZTU2JteFBZck45Q0ZxUms0SzRndyJ9.eyJhdWQiOiI1MDU2MDI0ZS1mYzZhLTQ5NWEtOWRkZC05MWQ0ZDVkMDhjOGQiLCJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubGluZS5jb20vZGZiYWRiOTAtYzA0Yy00NzBhLTkxZDctMzk3NzAyMDY4MTdiL3YyLjAiLCJpYXQiOjE2MDEzNzg2NzMsIm5iZiI6MTYwMTM3ODY3MywiZXhwIjoxNjAxMzgyNTczLCJhaW8iOiJBVFFBeS84UkFBQUE2MU9kdFphZTFTTDdNdVpFRjZPc2N4bmUyUE4wZ3pWbldMeGZlRUUvSWhBL2w5YUxQaWNLc0RldFVFQ2p3cjhMIiwicmgiOiIwLkFBQUFrTnU2MzB6QUNrZVIxemwzQWdhQmUwNENWbEJxX0ZwSm5kMlIxTlhRakkxMEFOQS4iLCJzdWIiOiJreHA4S1hRejJ6a3lBcHRYbnYzVzBndWxFcXJvaU00TkpyWFQ1N1VzLU5zIiwidGlkIjoiZGZiYWRiOTAtYzA0Yy00NzBhLTkxZDctMzk3NzAyMDY4MTdiIiwidXRpIjoiTV9wLXpPaTY5MGlvVFF5emJDYWhBQSIsInZlciI6IjIuMCJ9.FmVCO0krEnCmaEXNjM1TKq2JH0UB1yaTBQSYxEJ7gmVwccZirxC7YZ35KoV138gYlvsVXJjfIumrKthmXKi7_vuXR_m6Dl7sv8EgjASpTtsH-lPvlVKOA2wHQzFVxjTx5uskjxneMqxFU1MQSI1JG77PVr4QNki30SbFUenVif4a_uoUXzwbxsR3WhknNi77IXO_Z23DbDiZuhIYRKhG6hu3nbnaFeYiKpySk1V9tTa-gPj73gVDNwizFAvG-DX8zCGrRrBnknSlVrLohu5rLVkBrY1roGp6bMJmve3vY0RgtpeGYeodoO1244nOii5Czyc_XY_yehTOk4QRDPHSFg&access_token=eyJ0eXAiOiJKV1QiLCJub25jZSI6ImxmVGFUN0J6c0lEZkg2YmZteDYwUmN5NEFsS05maGEyT3FXMFExUUh1UmciLCJhbGciOiJSUzI1NiIsIng1dCI6ImppYk5ia0ZTU2JteFBZck45Q0ZxUms0SzRndyIsImtpZCI6ImppYk5ia0ZTU2JteFBZck45Q0ZxUms0SzRndyJ9.eyJhdWQiOiIwMDAwMDAwMy0wMDAwLTAwMDAtYzAwMC0wMDAwMDAwMDAwMDAiLCJpc3MiOiJodHRwczovL3N0cy53aW5kb3dzLm5ldC9kZmJhZGI5MC1jMDRjLTQ3MGEtOTFkNy0zOTc3MDIwNjgxN2IvIiwiaWF0IjoxNjAxMzc4NjczLCJuYmYiOjE2MDEzNzg2NzMsImV4cCI6MTYwMTM4MjU3MywiYWNjdCI6MCwiYWNyIjoiMSIsImFpbyI6IkUyUmdZUGh2N0RkYit2NmtEOXJYS3VlZjc2eUovM1pMYUdwU3MwOTFsUG5jbmtjYWJJSUEiLCJhbXIiOlsicHdkIl0sImFwcF9kaXNwbGF5bmFtZSI6Ik9TTUFGIEFkdmVudHVyZXMiLCJhcHBpZCI6IjUwNTYwMjRlLWZjNmEtNDk1YS05ZGRkLTkxZDRkNWQwOGM4ZCIsImFwcGlkYWNyIjoiMCIsImlkdHlwIjoidXNlciIsImlwYWRkciI6IjEwMy4xMjQuOTcuMiIsIm5hbWUiOiJUZXN0IFVzZXIiLCJvaWQiOiJiNmI2MjA0YS00MDQ5LTQwNmUtOWIxZi03YWQ5NzEzM2VjOTEiLCJwbGF0ZiI6IjEiLCJwdWlkIjoiMTAwMzIwMDBFNkM3RjUxMyIsInJoIjoiMC5BQUFBa051NjMwekFDa2VSMXpsM0FnYUJlMDRDVmxCcV9GcEpuZDJSMU5YUWpJMTBBTkEuIiwic2NwIjoiRGlyZWN0b3J5LkFjY2Vzc0FzVXNlci5BbGwgVXNlci5SZWFkIHByb2ZpbGUgb3BlbmlkIGVtYWlsIiwic3ViIjoia0FxNlBzSkVRcW0ybVRpdWlKOURDaWlkVGROQXhJLTRtb0duLWhjR2FRWSIsInRlbmFudF9yZWdpb25fc2NvcGUiOiJFVSIsInRpZCI6ImRmYmFkYjkwLWMwNGMtNDcwYS05MWQ3LTM5NzcwMjA2ODE3YiIsInVuaXF1ZV9uYW1lIjoidGVzdEBhZG1pbm9uZXNoZWVwLm9ubWljcm9zb2Z0LmNvbSIsInVwbiI6InRlc3RAYWRtaW5vbmVzaGVlcC5vbm1pY3Jvc29mdC5jb20iLCJ1dGkiOiJNX3Atek9pNjkwaW9UUXl6YkNhaEFBIiwidmVyIjoiMS4wIiwid2lkcyI6WyJiNzlmYmY0ZC0zZWY5LTQ2ODktODE0My03NmIxOTRlODU1MDkiXSwieG1zX3N0Ijp7InN1YiI6Imt4cDhLWFF6MnpreUFwdFhudjNXMGd1bEVxcm9pTTROSnJYVDU3VXMtTnMifSwieG1zX3RjZHQiOjE2MDA4NzM3NDB9.Zk6E0SfJIkaVEbv77CmDGDaCRTxPIBbfq60j4f7fIvi1aFbNtFAMj1uhUmcQ3BpH1aoYVpwbqyjlZmTRiiouQsBrYJCwI2QZ9SW3eZA5N9CxGcI8kgmSDypkhKc0R29u9mpZg-fLSJotHlfgetpMmV2NrybG6mFv2gl4H8mnZXDI50we63sZOFnICXuCXOv2Q0NVAN30FCji_EnpC7CH7FYu-j0lV6w1BDwsNNlQzYNFFf1UuixE4M9X1aBkKtb-l-jS07YG6tV0LJzZ5jKZAeNwJOBhP1S89S6uedWynHNBInpZqP8RhjsATCPdoWABQeLabA5zl8BhFP_ABbxBwQ&providerId=microsoft.com ]
E/flutter ( 5777): #0      catchPlatformException (package:firebase_auth_platform_interface/src/method_channel/utils/exception.dart:20:3)
E/flutter ( 5777): #1      _rootRunUnary (dart:async/zone.dart:1198:47)
E/flutter ( 5777): #2      _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter ( 5777): #3      _FutureListener.handleError (dart:async/future_impl.dart:160:20)
E/flutter ( 5777): #4      Future._propagateToListeners.handleError (dart:async/future_impl.dart:708:47)
E/flutter ( 5777): #5      Future._propagateToListeners (dart:async/future_impl.dart:729:24)
E/flutter ( 5777): #6      Future._completeError (dart:async/future_impl.dart:537:5)
E/flutter ( 5777): #7      _AsyncAwaitCompleter.completeError (dart:async-patch/async_patch.dart:47:15)
E/flutter ( 5777): #8      MethodChannel.invokeMapMethod (package:flutter/src/services/platform_channel.dart)
E/flutter ( 5777): <asynchronous suspension>
E/flutter ( 5777): #9      MethodChannelFirebaseAuth.signInWithCredential (package:firebase_auth_platform_interface/src/method_channel/method_channel_firebase_auth.dart:389:47)
E/flutter ( 5777): #10     FirebaseAuth.signInWithCredential (package:firebase_auth/src/firebase_auth.dart:436:31)
E/flutter ( 5777): #11     _AuthGuardState._loginWithMicrosoft (package:maf_adventures/main.dart:173:33)
E/flutter ( 5777): #12     _rootRunUnary (dart:async/zone.dart:1198:47)
E/flutter ( 5777): #13     _CustomZone.runUnary (dart:async/zone.dart:1100:19)
E/flutter ( 5777): #14     _FutureListener.handleValue (dart:async/future_impl.dart:143:18)
E/flutter ( 5777): #15     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:696:45)
E/flutter ( 5777): #16     Future._propagateToListeners (dart:async/future_impl.dart:725:32)
E/flutter ( 5777): #17     Future._completeWithValue (dart:async/future_impl.dart:529:5)
E/flutter ( 5777): #18     Future._asyncCompleteWithValue.<anonymous closure> (dart:async/future_impl.dart:567:7)
E/flutter ( 5777): #19     _rootRun (dart:async/zone.dart:1190:13)
E/flutter ( 5777): #20     _CustomZone.run (dart:async/zone.dart:1093:19)
E/flutter ( 5777): #21     _CustomZone.runGuarded (dart:async/zone.dart:997:7)
E/flutter ( 5777): #22     _CustomZone.bindCallbackGuarded.<anonymous closure> (dart:async/zone.dart:1037:23)
E/flutter ( 5777): #23     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
E/flutter ( 5777): #24     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
E/flutter ( 5777): 

```

I am not sure why I get this `INVALID_CREDENTIAL_OR_PROVIDER_ID` :(

**Reply by paurakhsharma** [reply]
One quick question, how did you find out that you can pass `tenent` in `customOAuthParameters`?
I couldn't find it in the docs of https://github.com/amrfarid140/firebase_auth_oauth.

I am asking because I want to know if I can pass the `redirect_uri` in it as well.

Because I am getting:
![image](https://user-images.githubusercontent.com/20378877/94557065-c8534500-027d-11eb-9882-659e6a88a0f5.png)

This error with this config: 
```
User user = await FirebaseAuthOAuth().openSignInFlow(
      "microsoft.com",
      ["email"],
      {
        'tenant': 'dfbadb90-c04c-470a-91d7-39770206817b',
      },
    );
```

Any idea why I might be getting this error?

**Reply by jthegedus** [reply]
>One quick question, how did you find out that you can pass `tenent` in `customOAuthParameters`?

I've been reading the source code of all these packages to try and understand how they differ and what they support. I generally open the repo in GitHub and use the search for particular terms. In this instance, `customOAuthParameters` is used the same here as in the other MSAL libs, so I just tried it.

>Any idea why I might be getting this error?

I think I saw that error when my AD redirect wasn't registered - https://firebase.google.com/docs/auth/android/microsoft-oauth#before_you_begin point 4.b.

>When registering apps with these providers, be sure to register the *.firebaseapp.com domain for your project as the redirect domain for your app (Azure AD app).

Or the tenant ID is incorrect for the Firebase AD Secret/AppID config in the Web Console

**Reply by jthegedus** [reply]
Another piece of documentation:

>With the Azure AD v2.0 endpoint, you specify the offline_access permission in the scope parameter to explicitly request a refresh token when using the OAuth 2.0 or OpenID Connect protocols

source: https://docs.microsoft.com/en-us/graph/permissions-reference#remarks-16

So perhaps you require both `openid` and `offline_access` to get the correct token with the AadOuath library?

**Reply by paurakhsharma** [reply]
@jthegedus Thank you so so much. [firebase_auth_oauth](https://github.com/amrfarid140/firebase_auth_oauth) method worked for me :tada: I hope you find success with `FirebaseAuth.instance.signInWithCredential(credential)` method.

**Reply by jthegedus** [reply]
I'm glad it helped 😄
